fix(tee): constant-time exact binomial sampler for IprfTee

igor53627 opened the pull request:
## Summary

Implements Issue #41: Replace arithmetic approximation with true Binomial(n,p) distribution for TEE constant-time execution.

## Changes

### Core Implementation
- **`binomial_sample_tee`**: True Binomial(n,p) using inverse-CDF with fixed O(n) iterations
- **`ct_f64_le`**: Constant-time IEEE 754 float comparison for [0,1) range
- **`binomial_inverse_ct`**: Fixed iteration loop with `ct_select_u64` (no early exit)

### TEE Hardening
- `IprfTee::new` asserts `n <= CT_BINOMIAL_MAX_COUNT` (4096) to prevent fallback
- `IprfTee::inverse_ct` uses CT min for MAX_PREIMAGES clamping
- Document `IprfTee::forward` is NOT constant-time (only inverse matters for PIR)

### Security Properties
- No branches depend on `prf_output` (secret-derived value)
- Always iterates exactly `count + 1` times
- Fallback for count > 4096 is unreachable in IprfTee due to constructor assertion

## Testing
- All 60 tests pass
- `test_tee_matches_standard_binomial`: Verifies TEE sampler matches non-TEE for small counts
- `test_tee_distribution_mean`: Validates true binomial distribution
- Proptest coverage for CT primitives

## Review
- Oracle reviewed and approved implementation
- CodeRabbit CLI found no issues

Closes #41

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds a constant-time exact Binomial(n,p) sampler and float CT comparisons, and updates IprfTee to use it with strict bounds and CT fixes.
> 
> - **Binomial sampling (`state-syncer/src/binomial.rs`)**:
>   - Add `binomial_sample_tee` (inverse-CDF, fixed-iteration, constant-time) with `CT_BINOMIAL_MAX_COUNT` (4096) and O(1) fallback for larger counts.
>   - Implement `binomial_inverse_ct` and TEE-only approximation helper.
>   - Extensive tests validating CT sampler vs standard, edge cases, distribution, and thresholds.
> - **Constant-time utilities (`state-syncer/src/constant_time.rs`)**:
>   - Add `ct_f64_le` and `ct_f64_lt` for IEEE-754 non-negative comparisons with tests/property checks.
> - **iPRF TEE integration (`state-syncer/src/iprf.rs`)**:
>   - `IprfTee::new` asserts `n <= CT_BINOMIAL_MAX_COUNT`.
>   - `trace_ball_inverse_ct` and `trace_ball` now use `binomial_sample_tee`.
>   - `inverse_ct` uses constant-time min for `MAX_PREIMAGES` clamping.
>   - Document `forward` is not constant-time (only inverse is).
> - **Changelog**:
>   - Document new TEE binomial sampler, CT float ops, and IprfTee changes.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 5f06cd50bce7646ce5d2ad0aaa2c51c93d4f4744. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * TEE-safe constant-time binomial sampler, constant-time float utilities, and full-domain PRP wrappers for strengthened, paper-faithful security.

* **Documentation**
  * Expanded changelog, security/TEE usage notes, and CLI/padding/hint documentation.

* **Tests**
  * Added unit/property tests and committed regression seeds for reproducible failures.

* **Chores**
  * CI/workflow improvements and changelog reorganization.

* **Formal**
  * Verification tooling, spec/proof adjustments, and build workflow updates.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

keewoolee commented on state-syncer/src/binomial.rs:
The variable `count` is not generally public; it is only public at the first recursion level of PMNS. We cannot assume that `count` is public or allow branching based on its value.

igor53627 commented:
@keewoolee

## Changes Implemented

All three review comments addressed:

### 1. `count` treated as secret (line 185 comment)
- Removed `if count == 0` branch, replaced with CT masking via `ct_eq_u64`
- Loop bounds now depend only on `CT_BINOMIAL_MAX_COUNT` (public constant), not on `n`
- All comparisons involving `n` use `ct_le_u64`, `ct_lt_u64`, `ct_select_*`

### 2. `CT_BINOMIAL_MAX_COUNT` increased to 65536 (line 184 comment)
- Covers Ethereum address use case (~40k at first recursion level)
- Added `test_tee_large_count_ethereum` test for n=40000

### 3. Approximation fallback removed entirely (line 264 comment)
- Deleted `binomial_sample_tee_approx` function
- TEE path always uses exact inverse-CDF via mode-based computation

### Numerical Stability
- Uses log-gamma (Lanczos) to compute PMF at mode
- Expands outward from mode to avoid underflow for large n
- 3 fixed-iteration passes, all O(CT_BINOMIAL_MAX_COUNT)

### Verification
- [OK] 75 tests pass (including new large-n test)
- [OK] Formal verification (Rocq proofs) pass
- [OK] Oracle review passed all 6 checkpoints
- [OK] CodeRabbit review passed

EOF

namnc commented on state-syncer/src/iprf.rs:
```diff
@@ -571,7 +571,20 @@ pub struct IprfTee {
+    /// sampling requires O(n) iterations; larger domains would either leak timing
+    /// or fall back to a non-Binomial approximation.
     pub fn new(key: PrfKey128, n: u64, m: u64) -> Self {
+        assert!(
```
- I think we can get rid of the panic in `pub fn binomial_sample_tee` and `fn binomial_inverse_ct` and just let the function accumulate up to count (add a second loop right after the main loop in case `count > CT_BINOMIAL_MAX_COUNT`, we won't have stablized proability accummulation but should be fine), this leakage should be acceptable and not causing the program to abort
```
    debug_assert!(
        count <= CT_BINOMIAL_MAX_COUNT,
        "binomial_sample_tee: count {} exceeds CT_BINOMIAL_MAX_COUNT {}",
        count,
        CT_BINOMIAL_MAX_COUNT
    );
```

namnc commented on state-syncer/src/binomial.rs:
- check if `let pmf_mode = log_pmf_mode.exp();` use some numerical trick to calculate (not ct)

igor53627 commented:
## Review Feedback Addressed

@namnc @keewoolee - All review comments from [your review](https://github.com/igor53627/plinko-extractor/pull/49#pullrequestreview-3581749175) have been addressed.

### Summary of Changes

| Issue | Change | Status |
|-------|--------|--------|
| #57 | CT helpers: `ct_min_u64`, `ct_max_u64`, `ct_saturating_sub_u64` | [OK] |
| #58 | Branchless `lgamma_approx` (removed dead `if x < 0.5` branch) | [OK] |
| #59 | Single-loop `binomial_inverse_ct` (**2.6x faster**) | [OK] |
| #60 | CT clamping for n > CT_BINOMIAL_MAX_COUNT | [OK] |
| #61 | Document exp() input range and CT considerations | [OK] |

Parent tracking issue: #56

### Oracle Verification: Changes Do NOT Break Canonical Plinko Specification

**What was checked:**
- `plinko/formal/specs/TrueBinomialSpec.v` - Inverse-CDF binomial specification
- `plinko/formal/specs/BinomialSpec.v` - Simplified arithmetic specification  
- `docs/Plinko.v` - Coq PMNS implementation
- `docs/plinko_paper_part6_algorithms.json` - Paper algorithms (HintInit, GetHint, etc.)

**Key findings:**
1. Formal specs constrain **behavior** (distribution, range, determinism), not **implementation** details
2. No Coq lemmas reference the Rust TEE implementation (`binomial_sample_tee`, `lgamma_approx`, etc.)
3. Paper treats `Binomial(n,p;r)` as a black-box oracle with no algorithmic requirements
4. All proven lemmas (`true_binomial_range`, `binom_quantile_le_n`, `binomial_sample_range`) remain valid

**Specific verification:**
- `TrueBinomialSpec.v` defines `true_binomial_sample_spec` via exact rational inverse-CDF - implementation-agnostic
- `BinomialSpec.v` defines the simpler `(count * num + (prf_output mod (denom + 1))) / denom` formula - different from TEE sampler
- `Plinko.v` PMNS uses `BinomialSpec` formula, not the inverse-CDF - no link to Rust code
- Paper `pmns_children` says "s <- Binomial(count, p; F(k, node))" with note "derandomized binomial sampling using r as randomness" - no implementation constraints

### Test Results

```
All tests pass:
- constant_time tests: 21 passed (including new CT helper proptests)
- binomial tests: 19 passed (including TEE tests)
- Test time: 18s (down from 47s - 2.6x faster)
```

Ready for re-review!

igor53627 commented:
@namnc Could you please resolve the review conversations once you've verified the changes? Thanks!
